field/TinyLang/Field/Typed/TypeChecker.hs

```diff
@@ -0,0 +1,382 @@
+typeCheck = runTypeChecker . inferExpr
+
+{-|
+-}
+checkType
+    :: forall f a. (TextField f, KnownUni f a)
+    => R.Expr R.Var f -> Either TypeCheckError (T.Expr f a)
+checkType = runTypeChecker . checkExpr
+
+
+{-| = Bidirectional Typing Rules
+
+Here is a quick recap of our current syntax of the language
+
+@
+identifiers         i
+variables           x       ::= i | ?i | #i
+boolean constants   B       ::= T | F
+field constants     F       ::= /determined by the choice of field/
+vector constants    V       ::= { B+ }
+integer constants   I
+constants           c       ::= B | F | V
+universes           U       ::= Bool | Field | Vector
+statements          S, T    ::= let x = M | assert M
+                              | for var = I to I do S* end
+expressions         L, M, N ::= c | x | L binOp M | unOp L | L [ M ] | S ; L
+                              | if L then M else N
+binary operations   binOp   ::= or | ...
+unary operations    unOp    ::= not | ...
+@
+
+We assume the followin metafunctions:
+
+* tyConst(c), which returns the type of a constant,
+
+* tyVar(x), which returns the type of a variable,
+
+* tyLeft(binOp), which returns the type of the first argument of binOp,
+
+* tyRight(binOp), which returns the type of the second argument of binOp,
+
+* tyResult(binOp), which returns the type of the result of binOp,
+
+* tyArg(unOp), which returns the type of the argument of unOp,
+
+* tyResult(unOp), which returns the type of the result of unOp.
+
+We split the type-checking into the following judgements:
+
+* Type checking judgement which checks the type of expression, for
+  which both the expression @M@ and the universe @U@ are inputs:
+  @ M^+ <= U^+ @
+
+* Type synthesis (inference) judgement which infers the type of the
+  expression, for which the expression @M@ is an input and the
+  universe @U@ is an output:
+  @ M^+ => U^- @
+
+* Type equality judgement which checks wether two universes match, for
+  which both the universe @U1@ and @U2@ are inputs:
+  @ U1^+ = U2^+ @
+
+We can switch from type checking judgement to type inference judgement
+using type equality
+
+@
+M => U2
+U2 = U1
+-------
+M <= U1
+@
+
+Please note that at the moment we lack type ascriptions and for
+convenience we add type inference rule for if-then-else expressions.
+
+-}
+
+mkSomeUniVar :: forall f. T.Var -> T.SomeUniVar f
+mkSomeUniVar var
+    | '?':_ <- _varName var = Some $ T.UniVar Bool   var
+    | '#':_ <- _varName var = Some $ T.UniVar Vector var
+    | otherwise             = Some $ T.UniVar Field  var
+
+{-| Type inference for variables
+-}
+inferUniVar
+    :: forall m f. (MonadSupply m, MonadScope m) => R.Var -> m (T.SomeUniVar f)
+inferUniVar = liftM mkSomeUniVar . makeVar . R.unVar
+
+{-| Type inference for expressions
+@
+M => U
+@
+-}
+inferExpr
+    :: forall m f. (MonadTypeChecker m, TextField f)
+    => R.Expr R.Var f -> m (T.SomeUniExpr f)
+{-|
+@
+---------------
+c => tyConst(c)
+@
+-}
```

Comment:
Just curious, are such comments rendered by Haddock?

Reply:
They can  be rendered nicely, if they are in a single comment block.

Unfortunately, they are not rendered very nicely when interleaving other function definitions.